TombEngine/Specific/Worker.cpp

```diff
@@ -0,0 +1,144 @@
+#include "framework.h"
+#include "Specific/Worker.h"
+
+namespace TEN::Utils
+{
+	WorkerManager g_Worker = WorkerManager();
+
+	WorkerManager::WorkerManager()
+	{
+		// Reserve threads.
+		unsigned int threadCount = std::thread::hardware_concurrency() * 2;
+		_threads.reserve(threadCount);
+
+		// Create threads.
+		for (int i = 0; i < threadCount; i++)
+			_threads.push_back(std::thread(&WorkerManager::Worker, this));
+
+		_deinitialize = false;
+	}
+
+	WorkerManager::~WorkerManager()
+	{
+		// LOCK: Restrict shutdown flag access.
+		{
+			auto taskLock = std::lock_guard(_taskMutex);
+
+			_deinitialize = true;
+		}
+
+		// Notify all threads they should stop.
+		_taskCond.notify_all();
+
+		// Join all threads.
+		for (auto& thread : _threads)
+		{
+			if (thread.joinable())
+				thread.join();
+		}
+	}
+
+	uint64_t WorkerManager::GetNewGroupId()
+	{
+		// Increment group ID counter, skipping NO_VALUE on wrap.
+		auto groupId = _groupIdCounter.fetch_add(1, std::memory_order_relaxed);
+		if (groupId == (uint64_t)NO_VALUE)
+			groupId = _groupIdCounter.fetch_add(1, std::memory_order_relaxed);
+
+		return groupId;
+	}
+
+	unsigned int WorkerManager::GetThreadCount() const
+	{
+		return (unsigned int)_threads.size();
+	}
+
+	unsigned int WorkerManager::GetCoreCount() const
+	{
+		return std::thread::hardware_concurrency();
+	}
+
+	void WorkerManager::AddTask(const WorkerTask& task, uint64_t groupId)
+	{
+		// LOCK: Restrict task queue access.
+		{
+			auto taskLock = std::lock_guard(_taskMutex);
+
+			// Add task to queue.
+			_tasks.push([this, task, groupId]() { HandleTask(task, groupId); });
+
+			// Increment group task count (if applicable).
+			if (groupId != (uint64_t)NO_VALUE)
+			{
+				// LOCK: Restrict group task count access.
+				{
+					auto groupLock = std::lock_guard(_groupMutex);
+
+					_groupTaskCounts[groupId]++;
+				}
+			}
+		}
+
+		// Notify one thread to handle task.
+		_taskCond.notify_one();
+	}
+
+	void WorkerManager::WaitForGroup(uint64_t groupId)
+	{
+		// LOCK: Restrict group task completion.
+		{
+			auto groupLock = std::unique_lock(_groupMutex);
+			_groupCond.wait(groupLock, [this, groupId]() { return (_groupTaskCounts.find(groupId) == _groupTaskCounts.end()); });
+		}
+	}
+
+	void WorkerManager::Worker()
+	{
+		while (true)
+		{
+			auto task = WorkerTask();
+
+			// LOCK: Restrict task queue access.
+			{
+				auto taskLock = std::unique_lock(_taskMutex);
+				_taskCond.wait(taskLock, [this] { return (_deinitialize || !_tasks.empty()); });
```

Comment:
Similar to what I suggested above, you should not have _definitialize be an atomic. Here, the conversion to bool through std::atomic<bool>::operator bool performs std::memory_order_seq_cst on the load operation, which is unnecessary strong considering you already locked a mutex when wait() woke up